politeiad/backendv2/tstorebe/plugins/pi/proposalstatus.go

```diff
@@ -0,0 +1,191 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package pi
+
+import (
+	"encoding/hex"
+
+	backend "github.com/decred/politeia/politeiad/backendv2"
+	"github.com/decred/politeia/politeiad/plugins/pi"
+	"github.com/decred/politeia/politeiad/plugins/ticketvote"
+)
+
+// getProposalStatus determines the proposal status in runtime, it uses the
+// in-memory cache to avoid retrieving the record, it's vote summary or
+// it's billing status changes when possible.
+func (p *piPlugin) getProposalStatus(token []byte) (pi.PropStatusT, error) {
+	var (
+		propStatus pi.PropStatusT
+		err        error
+		tokenStr   = hex.EncodeToString(token)
+
+		// The following fields are required to determine the proposal status and
+		// MUST be populated.
+		recordState  backend.StateT
+		recordStatus backend.StatusT
+		voteStatus   ticketvote.VoteStatusT
+
+		// The following fields are required to determine the proposal status and
+		// will only be populated for certain types of proposals or during certain
+		// stages of the proposal lifecycle.
+		voteMetadata    *ticketvote.VoteMetadata
+		billingStatuses []pi.BillingStatusChange
+	)
+
+	// Check if the proposal status has been cached
+	e := p.statuses.get(tokenStr)
+	if e != nil {
+		propStatus = e.propStatus
+		recordState = e.recordState
+		recordStatus = e.recordStatus
+		voteStatus = e.voteStatus
+		voteMetadata = e.voteMetadata
+	}
+
+	// Check if we need to get any additional data
+	if statusIsFinal(propStatus) {
+		// The status is final and cannot be changed.
+		// No need to get any additional data.
+		return propStatus, nil
+	}
+
+	// Get the record if required
+	if statusRequiresRecord(propStatus) {
+		r, err := p.record(backend.RecordRequest{
+			Token:     token,
+			Filenames: []string{ticketvote.FileNameVoteMetadata},
+		})
+		if err != nil {
+			return "", err
+		}
+
+		// Update the record data fields required to
+		// determine the proposal status.
+		recordState = r.RecordMetadata.State
+		recordStatus = r.RecordMetadata.Status
+		voteStatus = ticketvote.VoteStatusInvalid
+
+		// Pull the vote metadata out of the record files
+		voteMetadata, err = voteMetadataDecode(r.Files)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// Get the vote summary if required
+	if requiresVoteSummary(propStatus, recordState, recordStatus) {
```

Comment:
Why are `requiresVoteSummary` and `requiresBillingStatuses` not just going off of the status?